azure-iot-hub/azure/iot/hub/digital_twin_client.py

```diff
@@ -12,20 +12,38 @@ class DigitalTwinClient(object):
     based on top of the auto generated IotHub REST APIs
     """
 
-    def __init__(self, connection_string):
+    def __init__(self, connection_string=None, host=None, auth=None):
         """Initializer for a DigitalTwinClient.
 
         After a successful creation the class has been authenticated with IoTHub and
         it is ready to call the member APIs to communicate with IoTHub.
 
         :param str connection_string: The IoTHub connection string used to authenticate connection
-            with IoTHub.
+            with IoTHub if we are using connection_str authentication. Default value: None
+        :param str host: The Azure service url if we are using token credential authentication.
+            Default value: None
+        :param str auth: The Azure authentication object if we are using token credential authentication.
+            Default value: None
 
         :returns: Instance of the DigitalTwinClient object.
         :rtype: :class:`azure.iot.hub.DigitalTwinClient`
         """
-        self.auth = ConnectionStringAuthentication(connection_string)
-        self.protocol = protocol_client(self.auth, "https://" + self.auth["HostName"])
+        if connection_string:
+            self.auth = ConnectionStringAuthentication(connection_string)
+            self.protocol = protocol_client(self.auth, "https://" + self.auth["HostName"])
+        else:
+            self.auth = auth
+            self.protocol = protocol_client(self.auth, "https://" + host)
```

Comment:
>self.protocol = protocol_client(self.auth, "https://" + host) [](http://example.com/codeflow?start=12&length=61)

Since `ConnectionStringAuthentication` has the the host name  in `self.auth["HostName"]`, could you also put it into `AzureIdentityCredentialAdapter` in the same way?

This way code would be more symmetric and you wouldn't need `host` as a parameter to `__init__`

Reply:
It is by design. The AzureIdentiyCredentialAdapter has nothing to do with the hostName. We need to keep it separate,